### PR TITLE
add `scan_messages` for iterating over messages

### PR DESCRIPTION
**Pull request by 5kbpers**
Signed-off-by: 5kbpers <hustmh@gmail.com>

### CODE REVIEW

File: src/engine.rs [hunk 1]
```diff
@@ -224,14 +224,25 @@ where
         Ok(None)
     }
 
-    pub fn get_message_leq<S: Message>(&self, region_id: u64, key: &[u8]) -> Result<Option<S>> {
+    pub fn scan_message<S, C>(
+        &self,
+        region_id: u64,
+        start_key: &[u8],
+        end_key: &[u8],
```
Comment (tabokie) [inline src/engine.rs:231]: Option?

File: src/engine.rs [hunk 2]
```diff
@@ -224,14 +224,25 @@ where
...
+        reverse: bool,
+        callback: C,
+    ) -> Result<()>
+    where
+        S: Message,
+        C: FnMut(&[u8], S) -> Result<bool>,
```
[the remark this answers is not in the document]
Comment (5kbpers) [reply src/engine.rs:237]: Used for catching errors when something goes wrong on value scanned.

[the remark this answers is not in the document]
Comment (tabokie) [reply src/engine.rs:237]: The point is user shouldn't fake their own errors as `crate::Error`. Say you can get away with wrapping it into a `Error::Other` now, but what if the `Other` variant is removed in the future?

File: src/engine.rs [hunk 3]
```diff
@@ -224,6 +224,47 @@ where
...
+        let _t = StopWatch::new(&*ENGINE_READ_MESSAGE_DURATION_HISTOGRAM);
+        if let Some(memtable) = self.memtables.get(region_id) {
+            memtable
+                .read()
+                .scan_message(start_key, end_key, reverse, callback)?;
+        }
+        Ok(())
+    }
+
+    pub fn scan<C>(
```
Comment (tabokie) [inline src/engine.rs:248]:
```suggestion
    pub fn scan_raw_messages<C>(
```

**Review by tabokie — COMMENTED** [review]
Rest looks good.

**Review by tabokie — APPROVED** [review]
Oh, update CHANGELOG.md.

File: CHANGELOG.md [hunk 4]
```diff
@@ -9,6 +9,7 @@
 ### New Features
 
 * Add `PerfContext` which records detailed time breakdown of the write process to thread-local storage.
+* Add `Engine::scan_messages` and `Engine::scan_raw_messages` for iterating over written key-values.
```
Comment (tabokie) [inline CHANGELOG.md:12]: Better under "Public API Changes"

**Pull request merged**